src/main/java/io/cdap/plugin/ariba/source/AribaStructuredTransformer.java

```diff
@@ -0,0 +1,366 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.ariba.source;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
+import io.cdap.cdap.api.common.Bytes;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.format.UnexpectedFormatException;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.plugin.ariba.source.util.ResourceConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * This {@code AribaStructuredTransformer} contains the logic to
+ * convert Ariba record to {@code StructuredRecord}
+ */
+public class AribaStructuredTransformer {
+  private static final Logger LOG = LoggerFactory.getLogger(AribaStructuredTransformer.class);
+
+  public StructuredRecord readFields(JsonNode nativeRawRecord, Schema outputSchema) {
+    StructuredRecord.Builder recBuilder = StructuredRecord.builder(outputSchema);
+    List<Schema.Field> cdapFields = outputSchema.getFields();
+
+    cdapFields.forEach(field -> {
+      String fieldName = field.getName();
+      Object fieldValStr = getFieldNativeValue(nativeRawRecord, fieldName);
+      Schema childSchema = field.getSchema();
+
+//       eg.
+//        {
+//          "Owner": {
+//            "SourceSystem": "ASM",
+//            "UserId": "customersupportadmin",
+//            "PasswordAdapter": "PasswordAdapter"
+//            }
+//        }
+      if (nativeRawRecord.get(fieldName) != null &&
+        nativeRawRecord.get(fieldName).getNodeType().name()
+          .equalsIgnoreCase(ResourceConstants.OBJECT) && !childSchema.getType().isSimpleType()) {
+        fieldValStr = readFields((JsonNode) fieldValStr, childSchema.getUnionSchema(0));
+      }
+
+//       e.g.
+//       {
+//         "Suppliers":[{
+//            "Suppliers": {
+//              "SourceSystem": "ASM",
+//              "SupplierId": "",
+//              "SupplierLocationId": ""
+//              }
+//           }]
+//       }
+      if (nativeRawRecord.get(fieldName) != null &&
+        nativeRawRecord.get(fieldName).getNodeType().name()
+          .equalsIgnoreCase(ResourceConstants.ARRAY) && !childSchema.getType().isSimpleType()) {
+        fieldValStr = readInternalDeltaFeed(childSchema, (JsonNode) fieldValStr);
+      }
+
+//         "State": "Active",
+      if (fieldValStr != null) {
```

Comment:
are we handling the case where the field is nullable and the field value is null?

Reply:
Yes I have seen cases where this comes as null and we don't have to handle cases when it's null as calling "recBuilder.build()" method already has the handling of null value.
below code is from build() method of StructuedRecord class whenever the value is null.
"fields.put(fieldName, null)"